spring-cloud-gcp-starters/spring-cloud-gcp-starter-core/src/main/java/org/springframework/cloud/gcp/core/GCPProperties.java

```diff
@@ -27,7 +28,7 @@
 
 	private String projectId;
 
-	private String jsonKeyFile;
+	private Resource credentialKeyFile;
```

Comment:
@meltsufin Depends where are you creating that `MessageHandler`, this is what I'm trying to explain.

So let's get some basic scenarios perhaps

1: Your constructor is `PubSubMessageHandler(String projectId, GoogleCredentials credentials)

Now, let's say you are using the starter package on a Boot app:


```java
@Autowired 
private GcpProperties properties;

@Bean
  @ServiceActivator(inputChannel = "pubsubOutputChannel")
  public MessageHandler messageSender(
      GoogleCredentials credentials) throws IOException {
    PubSubMessageHandler outboundAdapter =
        new PubSubMessageHandler(properties.getProjectId(), credentials);
    outboundAdapter.setTopic("test");
    return outboundAdapter;
  }
```

That makes sense for a Boot app.

But what if your project is not a boot app?

```java
//note you could use @Value too
@Autowired
private Environment env;
  @Bean
  public GoogleCloudCredentials credentials(){
// create credentials manually here, somehow
}
  @Bean
  @ServiceActivator(inputChannel = "pubsubOutputChannel")
  public MessageHandler messageSender(GoogleCredentials credentials) throws IOException {
    PubSubMessageHandler outboundAdapter =
        new PubSubMessageHandler(env.getProperty("MY_ENV_PROPERTY_FOR_PROJECT_ID"), credentials);
    outboundAdapter.setTopic("test");
    return outboundAdapter;
  }
```

My main concern is explaining how you keep those two concerns separate (AutoConfiguration module along with the Property binding) and your MessageHandler code.

Hope this helps